src/wp-includes/post.php

```diff
@@ -7311,37 +7311,39 @@ function update_post_caches( &$posts, $post_type = 'post', $update_term_cache =
 
 	update_post_cache( $posts );
 
-	$post_ids = array();
-	foreach ( $posts as $post ) {
-		$post_ids[] = $post->ID;
+	if ( $update_term_cache ) {
+		update_post_term_caches( $posts, $post_type );
 	}
 
-	if ( ! $post_type ) {
-		$post_type = 'any';
+	if ( $update_meta_cache ) {
+		$post_ids = wp_list_pluck( $posts, 'ID' );
+		update_postmeta_cache( $post_ids );
 	}
+}
 
-	if ( $update_term_cache ) {
-		if ( is_array( $post_type ) ) {
-			$ptypes = $post_type;
-		} elseif ( 'any' === $post_type ) {
-			$ptypes = array();
-			// Just use the post_types in the supplied posts.
-			foreach ( $posts as $post ) {
-				$ptypes[] = $post->post_type;
-			}
-			$ptypes = array_unique( $ptypes );
-		} else {
-			$ptypes = array( $post_type );
-		}
+/**
+ * Updates the cache for the given list of post objects.
+ *
+ * @since 6.4.0
+ *
+ * @param WP_Post[]     $posts     Array of post objects.
+ * @param string|array  $post_type Optional. Post type or an array of post types. If 'any' is passed then post type is read from post objects. . Default 'any'.
+ */
+function update_post_term_caches( $posts, $post_type = 'any' ) {
```

Comment:
It would be great if you could add test coverage directly for this function, e.g. 3 tests, one for each alternative way to provide the `$post_type` parameter.

Reply:
Added unit tests.